pkg/trace/api/proxy.go

```diff
@@ -0,0 +1,192 @@
+	urls, apiKeys, err := pec.proxyEndpoints(apiKey)
+	return &proxyMultiTransport{
+		rt:      client,
+		targets: urls,
+		keys:    apiKeys,
+		baseTag: "type:" + pec.name,
+	}, err
+
+}
+func (pec *proxyEndpointsConfig) proxyEndpoints(apiKey string) (urls []*url.URL, apiKeys []string, err error) {
+	main := pec.defaultURL
+	if v := config.Datadog.GetString(pec.ddURLConfig); v != "" {
+		main = v
+	} else if site := config.Datadog.GetString("site"); site != "" {
+		main = fmt.Sprintf(pec.urlTemplate, site)
+	}
+	u, err := url.Parse(main)
+	if err != nil {
+		// if the main intake URL is invalid we don't use additional endpoints
+		return nil, nil, fmt.Errorf("error parsing main %s intake URL %s: %v", pec.name, main, err)
+	}
+	urls = append(urls, u)
+	apiKeys = append(apiKeys, apiKey)
+
+	if pec.additionalEndpointsConfig != "" && config.Datadog.IsSet(pec.additionalEndpointsConfig) {
+		extra := config.Datadog.GetStringMapStringSlice(pec.additionalEndpointsConfig)
+		for endpoint, keys := range extra {
+			u, err := url.Parse(endpoint)
+			if err != nil {
+				log.Errorf("Error parsing additional %s intake URL %s: %v", pec.name, endpoint, err)
+				continue
+			}
+			for _, key := range keys {
+				urls = append(urls, u)
+				apiKeys = append(apiKeys, key)
+			}
+		}
+	}
+	return urls, apiKeys, nil
+}
+
+// forwardingProxyHandler returns a new HTTP handler which will proxy requests to the configured intakes.
+// If the main intake URL can not be computed because of config, the returned handler will always
+// return http.StatusInternalServerError along with a clarification.
+func (r *HTTPReceiver) forwardingProxyHandler(pec *proxyEndpointsConfig) http.Handler {
+	if pec.enabledConfig != "" && !config.Datadog.GetBool(pec.enabledConfig) {
+		return pec.proxyDisabledHandler(nil)
+	}
+	transport, err := pec.proxyRoundTripper(r.conf.NewHTTPTransport(), r.conf.APIKey())
+	if err != nil {
+		return pec.proxyDisabledHandler(err)
+	}
+
+	return pec.newForwardingProxy(transport, r.conf.Hostname, r.conf.DefaultEnv)
+}
+
+func (pec *proxyEndpointsConfig) proxyDisabledHandler(err error) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if err != nil {
+			msg := fmt.Sprintf("%s proxy forwarder is OFF: %v", pec.name, err)
+			http.Error(w, msg, http.StatusInternalServerError)
+		} else {
+			msg := fmt.Sprintf("%s proxy forwarder is OFF", pec.name)
+			http.Error(w, msg, http.StatusMethodNotAllowed)
+		}
+	})
+}
+
+// multiTransport sends HTTP requests to multiple targets using an
+// underlying http.RoundTripper. API keys are set separately for each target.
+// The target hostname
+// When multiple endpoints are in use the response from the main endpoint
+// is proxied back to the client, while for all aditional endpoints the
+// response is discarded. There is no de-duplication done between endpoint
+// hosts or api keys.
+type proxyMultiTransport struct {
```

Comment:
I think I preferred the previous name much more. I don't see the point of adding the prefix `proxy*`.

```suggestion
type multiTransport struct {
```